File: app/src/main/java/io/ray/hexis/presenter/abs/IMatrixPresenter.java
package io.ray.hexis.presenter.abs;

import android.support.v4.view.ViewPager;

import io.ray.hexis.model.QuadrantItem;
import io.ray.hexis.model.abs.IMatrixModel;
import io.ray.hexis.model.abs.IQuadrantModel;
import io.ray.hexis.presenter.QuadrantFragmentPagerAdapter;
import io.ray.hexis.view.abs.IMatrixFragment;

import java.util.List;

/**
 * Presenter for an IMatrixFragment.
 */
public interface IMatrixPresenter extends IPresenter<IMatrixFragment, IMatrixModel> {

  /**
   * Construct a QuadrantItem and add it to the IQuadrantModel for a specified quadrant.
   * @param quadrant  Quadrant that QuadrantItem is being added to
   * @param message   Message which will be used to construct a QuadrantItem
   */
  void addItem(int quadrant, String message);

  void addItem(int quadrant, QuadrantItem item);

  /**
   * Return the IQuadrantModel for a specified quadrant.
   * @param quadrant  Quadrant that data is requested from
   * @return          IQuadrantModel for a quadrant
   */
  IQuadrantModel getQuadrantData(int quadrant);

  void setQuadrantData(int quadrant, List<QuadrantItem> data);

  ViewPager getPager();
}

File: app/src/main/java/io/ray/hexis/presenter/abs/IMatrixPresenter.kt
package io.ray.hexis.presenter.abs

import android.support.v4.view.ViewPager
import io.ray.hexis.model.QuadrantItem
import io.ray.hexis.model.abs.IMatrixModel
import io.ray.hexis.model.abs.IQuadrantModel
import io.ray.hexis.view.abs.IMatrixFragment

/**
 * Presenter for an IMatrixFragment.
 */
interface IMatrixPresenter : IPresenter<IMatrixFragment, IMatrixModel> {

  // ViewPager from the managed fragment
  val pager : ViewPager

  /**
   * Add a QuadrantItem to a specified quadrant.
   * @param quadrant    Specific quadrant to add the item
   * @param item        New item
   */
  fun addItem(quadrant: Int, item: QuadrantItem)

  /**
   * Return the IQuadrantModel for a specified quadrant.
   * @param quadrant  Quadrant that data is requested from
   * @return          IQuadrantModel for a quadrant
   */
  fun getQuadrantData(quadrant: Int): IQuadrantModel

  /**
   * Set the data set for a given Quadrant.
   * @param quadrant    Specific quadrant
   * @param data        New data set
   */
  fun setQuadrantData(quadrant: Int, data: List<QuadrantItem>)

  /**
   * Notify the DB that an item has been removed from the DB.
   */
  fun notifyItemRemoved(item: QuadrantItem)

  /**
   * Notify the DB that an item has been modified.
   * @param item      Modified QuadrantItem
   * @param quadrant  Specific quadrant
   */
  fun notifyItemModified(item: QuadrantItem, quadrant: Int)
}

